src/herder/TxQueueLimiter.cpp

```diff
@@ -10,244 +10,216 @@
-        {
-            auto minFee = computeBetterFee(newTx, (*it)->getFeeBid(),
-                                           (*it)->getNumOperations());
-            return std::make_pair(false, minFee);
-        }
-        // ensure that this transaction is not from the same account
-        auto& tx = *it;
-        if (tx->getSourceID() == id)
-        {
-            return std::make_pair(false, 0ll);
-        }
-        auto curOps = tx->getNumOperations();
-        if (neededOps <= curOps)
-        {
-            return std::make_pair(true, 0ll);
-        }
-        neededOps -= curOps;
+        resetTxs();
     }
 
-    // we reach this point if the queue doesn't have capacity for that
-    // transaction even when empty. Combination of multiplier and max ledger
-    // size is too small for whatever reason
-    static size_t lastMax = std::numeric_limits<size_t>::max();
-    if (lastMax != maxQueueSizeOps())
+    // If some transactions were evicted from this or generic lane, make sure
+    // that the new transaction is better (even if it fits otherwise). This
+    // guarantees that we don't replace transactions with higher bids with
+    // transactions with lower bids and less operations.
+    int64_t minFeeToBeatEvicted = std::max(
+        computeBetterFee(
+            mLaneEvictedFeeBid[mSurgePricingLaneConfig->getLaneClassifier()(
+                *newTx)],
+            *newTx),
+        computeBetterFee(
+            mLaneEvictedFeeBid[SurgePricingPriorityQueue::GENERIC_LANE],
+            *newTx));
+    if (minFeeToBeatEvicted > 0)
     {
-        lastMax = maxQueueSizeOps();
-        CLOG_WARNING(Herder,
-                     "Transaction Queue limiter configured with {} operations: "
-                     "node won't be able to accept all transactions",
-                     lastMax);
+        return std::make_pair(false, minFeeToBeatEvicted);
     }
 
-    return std::make_pair(false, 0ll);
-}
-
-TransactionFrameBasePtr
-TxQueueLimiter::getWorstTransaction()
-{
-    auto it = mTxs->begin();
-    if (it == mTxs->end())
+    uint32_t txOpsDiscount = 0;
+    if (oldTx)
     {
-        throw std::logic_error(
-            "invalid state getting worst tx in TxQueueLimiter");
+        auto newTxOps = newTx->getNumOperations();
+        auto oldTxOps = oldTx->getNumOperations();
+        // Bump transactions must have at least the old amount of operations.
+        releaseAssert(oldTxOps <= newTxOps);
+        txOpsDiscount = newTxOps - oldTxOps;
     }
-    return *it;
+    return mTxs->canFitWithEviction(*newTx, txOpsDiscount, mEvictionCache);
 }
 
-bool
+void
 TxQueueLimiter::evictTransactions(
-    size_t ops, std::function<void(TransactionFrameBasePtr const&)> evict)
+    TransactionFrameBaseConstPtr newTx,
+    std::function<void(TransactionFrameBasePtr const&)> evict)
 {
-    while (size() + ops > maxQueueSizeOps())
+    for (auto const& [evictedStack, evictedDueToLaneLimit] : mEvictionCache)
```

Comment:
did you intend to evict everything from this cache?
Old code allowed for the possibility that by kicking out 1 tx we actually evict multiple tx (because of dependencies tracked by `TransactionQueue`).

Reply:
this means that `mEvictionCache` may not even track transactions that are in `mTx` anymore

Reply:
(can't really tell which invariant are supposed to hold)